src/onepass.rs

```diff
@@ -0,0 +1,1187 @@
+                    // closed-open ranges.
+                    for byte in (inst.start as usize)..(inst.end as usize + 1) {
+                        let byte = byte as u8;
+                        let bc = self.prog.byte_classes[byte as usize];
+                        trans.0[bc as usize] = Transition {
+                            tgt: TransitionTarget::BytesInst(child_idx),
+                            priority: priority
+                        };
+                    }
+                    children.push(child_idx);
+                }
+                &Inst::Split(ref inst) => {
+                    resume.push(inst.goto1);
+                    resume.push(inst.goto2);
+                }
+                &Inst::Match(_) => {
+                    self.accepting_states[inst_idx] = true;
+                    for t in trans.0.iter_mut() {
+                        // Note that we go from lowest to highest
+                        // priority, so we don't have to worry about
+                        // clobbering higher priority transitions here.
+                        *t = Transition {
+                            tgt: TransitionTarget::Match,
+                            priority: priority
+                        };
+                    }
+                }
+                &Inst::Ranges(_) | &Inst::Char(_) => unreachable!(),
+            }
+            priority += 1;
+        }
+
+        self.transitions[inst_idx] = Some(trans);
+
+        Ok(children)
+    }
+
+    /// Execute a forwarding job.
+    ///
+    /// To make that a little more concrete, consider the program snippet:
+    ///
+    /// 0000: Bytes(a, a)
+    /// 0001: Save(2)
+    /// 0002: Bytes(b, b)
+    ///
+    /// Here the state for `Bytes(a, a)` needs to transition to
+    /// the state for `Save(2)`, but it does not know when to do
+    /// so. The right answer is that it should transition to
+    /// the `Save(2)` state when it sees a `b`, but it is hard
+    /// to know what children `Save(2)` has from where `Bytes(a, a)`
+    /// stands. To handle this we just emit a forwarding job
+    /// that says "when you know enough about the `Save(2)` state,
+    /// please forward `Bytes(a, a)` to `Save(2)`.". We need to use
+    /// a full DAG for this because there could be multiple forwarding
+    /// states in a row:
+    ///
+    /// 0000: Bytes(a, a)
+    /// 0001: Save(2)
+    /// 0002: Save(3)
+    /// 0003: Bytes(b, b)
+    ///
+    /// Here we will end up with two forwarding jobs:
+    ///
+    /// 1. Forward from `Bytes(a, a)` to `Save(2)`.
+    /// 2. Forward from `Save(2)` to `Save(3)`.
+    ///
+    /// Which we structure as a dag that looks like:
+    ///
+    /// (2) --> (1)
+    ///
+    /// The arrow flows in a funny direction because we want the jobs
+    /// with no dependencies to live at the roots of the DAG so that
+    /// we can process them first.
+    fn perform_forward(&mut self, fwd: Forward) {
```

Comment:
I didn't see this whole forwarding scheme in the RE2 code, and I think that the reason boils down to the fact that there is DFA states in the RE2 encoding occupy a fixed amount of space in the transition table, so you can point to a state that has not yet been computed pretty easily. To keep the DFA state encoding small, RE2 devotes more of its equivalent of StatePointers to flags and control. One result of this is that RE2's onepass engine only supports a finite number of capture groups.

It might be worth trying to improve the OnePassCompiler performance by simplifying our DFA encoding a bit, but I think that would probably fit better in a separate PR.

Reply:
Another concern with forwarding is the time complexity. The graph algorithms involved are O(N*E) where N is the number of instructions and E is the edges between those instructions. It shouldn't be bad if the number of edges is O(N), which feels like it ought to be true, but I don't have an off the cuff proof about why. Its at least worth thinking about.